src/persist-client/src/internal/maintenance.rs

```diff
@@ -30,6 +32,53 @@ pub struct LeaseExpiration {
     pub(crate) writers: Vec<WriterId>,
 }
 
+impl LeaseExpiration {
+    pub(crate) async fn perform<K, V, T, D>(
+        self,
+        machine: &Machine<K, V, T, D>,
+    ) -> RoutineMaintenance
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let expire_readers = self.readers.into_iter().map(|expired| {
+            let mut machine = machine.clone();
+            mz_ore::task::spawn(|| "persist::automatic_read_expiration", async move {
+                info!(
+                    "Force expiring reader ({}) of shard ({}) due to inactivity",
+                    expired,
+                    machine.shard_id()
+                );
+                // XXX This ought to return maintenance too... along with most of the other
+                // explicit ::default() calls in this file.
+                let _ = machine.expire_leased_reader(&expired).await;
+                RoutineMaintenance::default()
+            })
+        });
+
+        let expire_writers = self.writers.into_iter().map(|expired| {
+            let mut machine = machine.clone();
+            mz_ore::task::spawn(|| "persist::automatic_write_expiration", async move {
+                info!(
+                    "Force expiring writer ({}) of shard ({}) due to inactivity",
+                    expired,
+                    machine.shard_id()
+                );
+                machine.expire_writer(&expired).await;
+                RoutineMaintenance::default()
+            })
+        });
+        let unordered: FuturesUnordered<_> = expire_readers.chain(expire_writers).collect();
+        // XXX again with the maintenance merge
+        // I think it's fine that we're not ordering these.
```

Comment:
we should be able to do some intelligent merges for each of the types of maintenance tasks, e.g. GC [already has de-duping logic](https://github.com/MaterializeInc/materialize/blob/19204780182a323b50fdba7a4948b742958500d9/src/persist-client/src/internal/gc.rs#L114-L137) which we might be able to reuse, lease expiry can take the union of all IDs passed in, rollups can use the latest seqno found (I think)

Reply:
Yeah, that makes sense -- will have a look at the existing dedup.